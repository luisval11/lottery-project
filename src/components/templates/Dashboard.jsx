const Dashboard = ({
	headerTitle,
	isOpenSort,
	items,
	selectedItem,
	sort,
	sortOptions,
	sortTitle,
	handleSelectItem,
	handleShowSortMenu,
	handleSelectOnChange,
}) => {
	return (
		<div class="dashboard">
			{headerTitle && <h4 className="dashboard__header">{headerTitle}</h4>}
		</div>
	);
};

export default Dashboard;
