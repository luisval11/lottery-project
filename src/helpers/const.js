export const URL_LOTTOLAND = "https://www.lottoland.co.uk";
export const SORT_OPTION_NAME = "name";
export const SORT_OPTION_EUR_STAKE_ASC = "stakeEurAsc";
export const SORT_OPTION_EUR_STAKE_DESC = "stakeEurDesc";
export const MAX_GAMES_PER_PAGE = 9;
export const SORT_OPTIONS = [
	{ name: SORT_OPTION_NAME, text: "Name (A-Z)" },
	{ name: SORT_OPTION_EUR_STAKE_ASC, text: "Stake (min to max)" },
	{ name: SORT_OPTION_EUR_STAKE_DESC, text: "Stake (max to min)" },
];
export const EXAMPLE_LISTITEM = {
	LuckyWizard: {
		name: "LuckyWizard",
		displayName: "Lucky Wizard",
		playURL: "/games/luckywizard/play",
		image: "/cms/5bab9d950eb3580fac83392e/Icon_320and250_LuckyWizard.jpg",
		detailURL: "/games/luckywizard",
		currencyData: {
			CHF: {
				noDesk: true,
				noMob: true,
			},
			EUR: {
				minimumStake: 0.1,
			},
			GBP: {
				minimumStake: 0.1,
			},
			SEK: { minimumStake: 1.0 },
			PLN: {},
			JPY: {},
			INR: {},
			NZD: { minimumStake: 0.1 },
			CAD: {},
			HUF: {},
		},
		volatility: ["High"],
		provider: ["RedTiger"],
	},
	GreekGods: {
		name: "GreekGods",
		displayName: "Greek Gods",
		playURL: "/games/greekgods/play",
		image: "/cms/5de91d2a91253e002efe72f7/icon-320and250_GreekGods.jpg",
		detailURL: "/games/greekgods",
		currencyData: {
			EUR: { minimumStake: 0.25 },
			GBP: { minimumStake: 0.25 },
			SEK: { minimumStake: 2.5 },
			PLN: {},
			BRL: {},
			JPY: {},
			NZD: { minimumStake: 0.25 },
			INR: {},
			HUF: {},
		},
		volatility: ["High"],
		provider: ["Pragmatic"],
	},
};
