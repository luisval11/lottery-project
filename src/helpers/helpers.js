import { SORT_OPTION_EUR_STAKE_ASC, SORT_OPTION_EUR_STAKE_DESC, SORT_OPTION_NAME, URL_LOTTOLAND } from "./const";

export const normalizeUrl = (endpoint) => {
	return URL_LOTTOLAND + endpoint;
};

export const isValidGame = (game) => {
	return (
		game.currencyData != undefined &&
		game.currencyData.EUR != undefined &&
		game.currencyData.EUR.minimumStake != undefined
	);
};

export const getItemsToShow = (sort, initialValues, initial, size) => {
	switch (sort) {
		case SORT_OPTION_NAME:
			return Object.values(initialValues)
				.filter((game) => isValidGame(game))
				.sort((gameA, gameB) => {
					return gameA.displayName.localeCompare(gameB.displayName);
				})
				.slice(initial, size);

		case SORT_OPTION_EUR_STAKE_ASC:
			return Object.values(initialValues)
				.filter((game) => isValidGame(game))
				.sort((gameA, gameB) => {
					return (
						gameA.currencyData.EUR.minimumStake -
						gameB.currencyData.EUR.minimumStake
					);
				})
				.slice(initial, size);

		case SORT_OPTION_EUR_STAKE_DESC:
			return Object.values(initialValues)
				.filter((game) => isValidGame(game))
				.sort((gameA, gameB) => {
					return (
						gameB.currencyData.EUR.minimumStake -
						gameA.currencyData.EUR.minimumStake
					);
				})
				.slice(initial, size);
	}
};
