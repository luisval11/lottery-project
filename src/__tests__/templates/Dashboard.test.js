import { h } from "preact";
import { mount, shallow } from "enzyme";
import { EXAMPLE_LISTITEM, SORT_OPTIONS } from "../../helpers/const";
import Dashboard from "../../components/templates/Dashboard";

describe("Test on Dashboard", () => {
	const headerTitle = "Header Title";
	const sortTitle = "Sort Title";
	let sort = "name";
	let isOpenSort = false;
	const items = [EXAMPLE_LISTITEM.LuckyWizard, EXAMPLE_LISTITEM.GreekGods];
	let selectedItem = null;
	const handleSelectItem = (value) => {
		selectedItem = value;
	};
	const handleShowSortMenu = () => {
		isOpenSort = !isOpenSort;
	};
	const handleSelectOnChange = (value) => {
		sort = value;
	};

	test("should match initial snapshot", () => {
		let wrapper = shallow(
			<Dashboard
				headerTitle={headerTitle}
				isOpenSort={isOpenSort}
				items={items}
				selectedItem={selectedItem}
				sort={sort}
				sortOptions={SORT_OPTIONS}
				sortTitle={sortTitle}
				handleSelectItem={handleSelectItem}
				handleShowSortMenu={handleShowSortMenu}
				handleSelectOnChange={handleSelectOnChange}
			/>
		);
		expect(wrapper).toMatchSnapshot();
	});

	test("should have a heading for the Dashboard", () => {
		let wrapper = mount(
			<Dashboard
				headerTitle={headerTitle}
				isOpenSort={isOpenSort}
				items={items}
				selectedItem={selectedItem}
				sort={sort}
				sortOptions={SORT_OPTIONS}
				sortTitle={sortTitle}
				handleSelectItem={handleSelectItem}
				handleShowSortMenu={handleShowSortMenu}
				handleSelectOnChange={handleSelectOnChange}
			/>
		);
		expect(wrapper.find("h4.dashboard__header").text()).toBe(headerTitle);
	});

	test("should have 1 Sort menu to sort the items on the List organism", () => {
		let wrapper = mount(
			<Dashboard
				headerTitle={headerTitle}
				isOpenSort={isOpenSort}
				items={items}
				selectedItem={selectedItem}
				sort={sort}
				sortOptions={SORT_OPTIONS}
				sortTitle={sortTitle}
				handleSelectItem={handleSelectItem}
				handleShowSortMenu={handleShowSortMenu}
				handleSelectOnChange={handleSelectOnChange}
			/>
		);
		expect(wrapper.find("div.sort__wrapper-select").length).toBe(1);
	});

	test("should have a separator to separate Sort organism and List organism", () => {
		let wrapper = mount(
			<Dashboard
				headerTitle={headerTitle}
				isOpenSort={isOpenSort}
				items={items}
				selectedItem={selectedItem}
				sort={sort}
				sortOptions={SORT_OPTIONS}
				sortTitle={sortTitle}
				handleSelectItem={handleSelectItem}
				handleShowSortMenu={handleShowSortMenu}
				handleSelectOnChange={handleSelectOnChange}
			/>
		);
		expect(wrapper.find("div.dashboard__rectangle").length).toBe(1);
	});

	test("should have 1 List organism to show the List of items according to the items received", () => {
		let wrapper = mount(
			<Dashboard
				headerTitle={headerTitle}
				isOpenSort={isOpenSort}
				items={items}
				selectedItem={selectedItem}
				sort={sort}
				sortOptions={SORT_OPTIONS}
				sortTitle={sortTitle}
				handleSelectItem={handleSelectItem}
				handleShowSortMenu={handleShowSortMenu}
				handleSelectOnChange={handleSelectOnChange}
			/>
		);
		expect(wrapper.find("div.dashboard__list").length).toBe(1);
	});

	test("if first 'sort' option is selected by default, when a different option is clicked, then first listItem is not 'selected' any more", () => {
		const optionSelected = 1;
		let wrapper = mount(
			<Dashboard
				headerTitle={headerTitle}
				isOpenSort={isOpenSort}
				items={items}
				selectedItem={selectedItem}
				sort={sort}
				sortOptions={SORT_OPTIONS}
				sortTitle={sortTitle}
				handleSelectItem={handleSelectItem}
				handleShowSortMenu={handleShowSortMenu}
				handleSelectOnChange={handleSelectOnChange}
			/>
		);

		//Extract the onChange for the 2nd input, as "name" is default sorting
		let onChange;
		let value;
		wrapper.find("input").forEach((input, id) => {
			if (id === optionSelected) {
				onChange = input.props().onChange;
				value = input.props().value;
			}
		});
		onChange({ target: { value: value } });

		//Forces a rerender of the component with the new value after onClick;
		wrapper.setProps({
			headerTitle: headerTitle,
			isOpenSort: isOpenSort,
			items: items,
			selectedItem: selectedItem,
			sort: value,
			sortOptions: SORT_OPTIONS,
			sortTitle: sortTitle,
			handleSelectItem: handleSelectItem,
			handleShowSortMenu: handleShowSortMenu,
			handleSelectOnChange: handleSelectOnChange,
		});
		wrapper.update();

		//Extract the onClick for the div input
		const { sort: sortProp } = wrapper.props();
		expect(sortProp).toBe(SORT_OPTIONS[optionSelected].name);
	});
});
