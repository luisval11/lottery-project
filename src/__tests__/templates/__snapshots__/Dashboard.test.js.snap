// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test on Dashboard should match initial snapshot 1`] = `ShallowWrapper {}`;
